File: matiasmasso/Mat.NET/DTO/DTOLocation.cs
using System;
using System.Collections.Generic;

namespace DTO
{
    public class DTOLocation : DTOArea
    {
        public DTOZona Zona { get; set; }
        public bool Export { get; set; }
        public DTOComarca Comarca { get; set; }
        public List<DTOZip> Zips { get; set; }
        public List<DTOContact> Contacts { get; set; }

        public DTOPortsCondicio PortsCondition { get; set; }

        public enum Wellknowns
        {
            madrid
        }

        public DTOLocation() : base()
        {
            base.Cod = Cods.Location;
            this.Zips = new List<DTOZip>();
        }

        public DTOLocation(Guid oGuid) : base(oGuid)
        {
            base.Cod = Cods.Location;
            this.Zips = new List<DTOZip>();
        }

        public static DTOLocation Factory(DTOZona zona, string nom = "")
        {
            DTOLocation retval = new DTOLocation();
            retval.Zona = zona;
            retval.Nom = nom;
            return retval;
        }

        public static DTOLocation Wellknown(DTOLocation.Wellknowns id)
        {
            DTOLocation retval = null;
            switch (id)
            {
                case DTOLocation.Wellknowns.madrid:
                    {
                        retval = new DTOLocation(new Guid("3EC9E266-C7D7-42B0-94DD-27334A5EE118"));
                        break;
                    }
            }
            return retval;
        }

        public static DTOLocation Factory(DTOZona oZona)
        {
            DTOLocation retval = new DTOLocation();
            retval.Zona = oZona;
            return retval;
        }

        public static string FullNom(DTOLocation oLocation, DTOLang oLang = null/* TODO Change to default(_) if this is not a reference type */)
        {
            string retval = "";
            if (oLocation != null)
                retval = oLocation.FullNom(oLang);
            return retval;
        }

        public string FullNom(DTOLang oLang = null/* TODO Change to default(_) if this is not a reference type */)
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            sb.Append(base.Nom);
            DTOZona oZona = this.Zona;
            if (oZona != null)
            {
                DTOCountry oCountry = this.Zona.Country;
                if (DTOArea.isEsp(oCountry))
                {
                    DTOAreaProvincia oProvincia = oZona.Provincia;
                    if (oProvincia == null)
                    {
                        if (base.Nom != oZona.Nom)
                            sb.Append(" (" + oZona.Nom + ")");
                    }
                    else if (base.Nom != oProvincia.Nom)
                        sb.Append(" (" + oProvincia.Nom + ")");
                }
                else if (oCountry != null)
                {
                    if (oLang == null)
                        oLang = DTOLang.ESP();
                    sb.Append(" (" + oCountry.LangNom.Tradueix(oLang) + ")");
                }
            }

            string retval = sb.ToString();
            return retval;
        }

        public static string FullNomSegmented(DTOLocation oLocation, DTOLang oLang)
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            sb.Append(oLocation.Nom);
            sb.Append("/");
            sb.Append(oLocation.Zona.FullNomSegmented(oLang));
            string retval = sb.ToString();
            return retval;
        }

        public static string FullNomSegmentedReversed(DTOLocation oLocation, DTOLang oLang)
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            sb.Append(oLocation.Zona.FullNomSegmented(oLang));
            sb.Append("/");
            sb.Append(oLocation.Nom);
            string retval = sb.ToString();
            return retval;
        }

        public string CountryNom(DTOLang oLang)
        {
            string retval = "";
            if (this.Zona != null)
                retval = this.Zona.CountryNom(oLang);
            return retval;
        }

        public bool Matches(string searchKey)
        {
            return MatHelperStd.TextHelper.Match(this.Nom, searchKey);
        }


        public static DTOInvoice.ExportCods ExportCod(DTOLocation oLocation)
        {
            DTOInvoice.ExportCods retval = DTOInvoice.ExportCods.notSet;
            if (oLocation != null)
            {
                if (oLocation.Zona != null)
                {
                    switch (oLocation.Zona.ExportCod)
                    {
                        case DTOInvoice.ExportCods.intracomunitari:
                        case DTOInvoice.ExportCods.extracomunitari:
                            {
                                retval = oLocation.Zona.ExportCod;
                                break;
                            }

                        default:
                            {
                                if (oLocation.Zona.Country != null)
                                    retval = oLocation.Zona.Country.ExportCod;
                                break;
                            }
                    }
                }
            }
            return retval;
        }

        public static DTOAreaProvincia Provincia(DTOLocation oLocation)
        {
            DTOAreaProvincia retval = null/* TODO Change to default(_) if this is not a reference type */;
            if (oLocation != null)
            {
                if (oLocation.Zona != null)
                    retval = oLocation.Zona.Provincia;
            }
            return retval;
        }

        public static string ProvinciaOrZonaNom(DTOLocation oLocation)
        {
            string retval = "";
            if (oLocation != null && oLocation.Zona != null)
            {
                if (oLocation.Zona.Provincia == null)
                    retval = oLocation.Zona.Nom;
                else
                    retval = oLocation.Zona.Provincia.Nom;
            }
            return retval;
        }
    }
}

File: matiasmasso/Mat.NET/DT2/DTOLocation.vb
Public Class DTOLocation
    Inherits DTOArea

    Shadows Property zona As DTOZona
    Property export As Boolean
    Property comarca As DTOComarca
    Property zips As List(Of DTOZip)
    Property contacts As List(Of DTOContact)

    Property portsCondition As DTOPortsCondicio

    Public Enum wellknowns
        madrid
    End Enum

    Public Sub New()
        MyBase.New()
        MyBase.Cod = Cods.Location
        _Zips = New List(Of DTOZip)
    End Sub

    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
        MyBase.Cod = Cods.Location
        _Zips = New List(Of DTOZip)
    End Sub

    Shared Function wellknown(id As DTOLocation.wellknowns) As DTOLocation
        Dim retval As DTOLocation = Nothing
        Select Case id
            Case DTOLocation.wellknowns.madrid
                retval = New DTOLocation(New Guid("3EC9E266-C7D7-42B0-94DD-27334A5EE118"))
        End Select
        Return retval
    End Function

    Overloads Shared Function Factory(oZona As DTOZona) As DTOLocation
        Dim retval As New DTOLocation
        retval.zona = oZona
        Return retval
    End Function

    Shared Shadows Function FullNom(oLocation As DTOLocation, Optional oLang As DTOLang = Nothing) As String
        Dim retval As String = ""
        If oLocation IsNot Nothing Then
            retval = oLocation.FullNom(oLang)
        End If
        Return retval
    End Function

    Public Shadows Function FullNom(Optional oLang As DTOLang = Nothing) As String
        Dim sb As New System.Text.StringBuilder
        sb.Append(MyBase.Nom)
        Dim oZona As DTOZona = _Zona
        If oZona IsNot Nothing Then
            Dim oCountry As DTOCountry = _Zona.Country
            If DTOArea.IsEsp(oCountry) Then
                Dim oProvincia As DTOAreaProvincia = oZona.Provincia
                If oProvincia Is Nothing Then
                    If MyBase.Nom <> oZona.Nom Then
                        sb.Append(" (" & oZona.Nom & ")")
                    End If
                Else
                    If MyBase.Nom <> oProvincia.Nom Then
                        sb.Append(" (" & oProvincia.Nom & ")")
                    End If
                End If
            Else
                If oCountry IsNot Nothing Then
                    If oLang Is Nothing Then oLang = DTOLang.ESP
                    sb.Append(" (" & oCountry.langNom.tradueix(oLang) & ")")
                End If
            End If
        End If

        Dim retval As String = sb.ToString
        Return retval
    End Function

    Shared Shadows Function FullNomSegmented(oLocation As DTOLocation, oLang As DTOLang) As String
        Dim sb As New System.Text.StringBuilder
        sb.Append(oLocation.Nom)
        sb.Append("/")
        sb.Append(oLocation.Zona.FullNomSegmented(oLang))
        Dim retval As String = sb.ToString
        Return retval
    End Function

    Shared Shadows Function FullNomSegmentedReversed(oLocation As DTOLocation, oLang As DTOLang) As String
        Dim sb As New System.Text.StringBuilder
        sb.Append(oLocation.Zona.FullNomSegmented(oLang))
        sb.Append("/")
        sb.Append(oLocation.Nom)
        Dim retval As String = sb.ToString
        Return retval
    End Function

    Public Function CountryNom(oLang As DTOLang) As String
        Dim retval As String = ""
        If _Zona IsNot Nothing Then
            retval = _Zona.CountryNom(oLang)
        End If
        Return retval
    End Function

    Shared Function ExportCod(oLocation As DTOLocation) As DTOInvoice.ExportCods
        Dim retval As DTOInvoice.ExportCods = DTOInvoice.ExportCods.NotSet
        If oLocation IsNot Nothing Then
            If oLocation.Zona IsNot Nothing Then
                Select Case oLocation.Zona.ExportCod
                    Case DTOInvoice.ExportCods.Intracomunitari, DTOInvoice.ExportCods.Extracomunitari
                        retval = oLocation.Zona.ExportCod
                    Case Else
                        If oLocation.Zona.Country IsNot Nothing Then
                            retval = oLocation.Zona.Country.ExportCod
                        End If
                End Select
            End If
        End If
        Return retval
    End Function

    Shared Shadows Function Provincia(oLocation As DTOLocation) As DTOAreaProvincia
        Dim retval As DTOAreaProvincia = Nothing
        If oLocation IsNot Nothing Then
            If oLocation.Zona IsNot Nothing Then
                retval = oLocation.Zona.Provincia
            End If
        End If
        Return retval
    End Function

    Shared Function ProvinciaOrZonaNom(oLocation As DTOLocation) As String
        Dim retval As String = ""
        If oLocation IsNot Nothing AndAlso oLocation.Zona IsNot Nothing Then
            If oLocation.Zona.Provincia Is Nothing Then
                retval = oLocation.Zona.Nom
            Else
                retval = oLocation.Zona.Provincia.Nom
            End If
        End If
        Return retval
    End Function
End Class
